include/boost/beast/websocket/detail/mask.hpp

```diff
@@ -83,181 +84,59 @@ using maskgen = maskgen_t<std::minstd_rand>;
 }
 
-// 32-bit optimized
-//
-template<class = void>
-void
-mask_inplace_fast(
-    boost::asio::mutable_buffer const& b,
-        std::uint32_t& key)
-{
-    auto n = b.size();
-    auto p = reinterpret_cast<std::uint8_t*>(b.data());
-    if(n >= sizeof(key))
-    {
-        // Bring p to 4-byte alignment
-        auto const i = reinterpret_cast<
-            std::uintptr_t>(p) & (sizeof(key)-1);
-        switch(i)
-        {
-        case 1: p[2] ^= static_cast<std::uint8_t>(key >> 16); BOOST_FALLTHROUGH;
-        case 2: p[1] ^= static_cast<std::uint8_t>(key >> 8);  BOOST_FALLTHROUGH;
-        case 3: p[0] ^= static_cast<std::uint8_t>(key);
-        {
-            auto const d = static_cast<unsigned>(sizeof(key) - i);
-            key = ror(key, 8*d);
-            n -= d;
-            p += d;
-            BOOST_FALLTHROUGH;
-        }
-        default:
-            break;
-        }
-    }
-
-    // Mask 4 bytes at a time
-    for(auto i = n / sizeof(key); i; --i)
-    {
-        *reinterpret_cast<
-            std::uint32_t*>(p) ^= key;
-        p += sizeof(key);
-    }
-
-    // Leftovers
-    n &= sizeof(key)-1;
-    switch(n)
-    {
-    case 3: p[2] ^= static_cast<std::uint8_t>(key >> 16); BOOST_FALLTHROUGH;
-    case 2: p[1] ^= static_cast<std::uint8_t>(key >> 8);  BOOST_FALLTHROUGH;
-    case 1: p[0] ^= static_cast<std::uint8_t>(key);
-        key = ror(key, static_cast<unsigned>(8*n));
-        BOOST_FALLTHROUGH;
-    default:
-        break;
-    }
-}
-
-// 64-bit optimized
+// Apply mask in place
 //
-template<class = void>
+inline
 void
-mask_inplace_fast(
-    boost::asio::mutable_buffer const& b,
-        std::uint64_t& key)
+mask_inplace(boost::asio::mutable_buffer& b, prepared_key& key)
 {
     auto n = b.size();
-    auto p = reinterpret_cast<std::uint8_t*>(b.data());
-    if(n >= sizeof(key))
+    auto mask = key; // avoid aliasing
+    auto p = reinterpret_cast<unsigned char*>(b.data());
+    while(n >= 4)
     {
-        // Bring p to 8-byte alignment
-        auto const i = reinterpret_cast<
-            std::uintptr_t>(p) & (sizeof(key)-1);
-        switch(i)
-        {
-        case 1: p[6] ^= static_cast<std::uint8_t>(key >> 48); BOOST_FALLTHROUGH;
-        case 2: p[5] ^= static_cast<std::uint8_t>(key >> 40); BOOST_FALLTHROUGH;
-        case 3: p[4] ^= static_cast<std::uint8_t>(key >> 32); BOOST_FALLTHROUGH;
-        case 4: p[3] ^= static_cast<std::uint8_t>(key >> 24); BOOST_FALLTHROUGH;
-        case 5: p[2] ^= static_cast<std::uint8_t>(key >> 16); BOOST_FALLTHROUGH;
-        case 6: p[1] ^= static_cast<std::uint8_t>(key >> 8);  BOOST_FALLTHROUGH;
-        case 7: p[0] ^= static_cast<std::uint8_t>(key);
-        {
-            auto const d = static_cast<
-                unsigned>(sizeof(key) - i);
-            key = ror(key, 8*d);
-            n -= d;
-            p += d;
-            BOOST_FALLTHROUGH;
-        }
-        default:
-            break;
-        }
+        for(int i = 0; i < 4; ++i)
```

Comment:
@vinniefalco,

I quite like this approach. It is clean, readable, and expresses the developer intent. Just a question though:

Does this generate equally efficient x86_64 code when the input buffer is not aligned at the start?

Reply:
You can thank Peter Dimov for this code, and the generated assembly is amazing. Alignment is irrelevant since it uses **movdqu**. You can see the results for various compilers here: https://godbolt.org/g/eNsgnY

Reply:
Nice! The mips compiler and architecture likely is not that smart, but I don't care too much.